packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -189,38 +195,55 @@ function useFocusContainment(scopeRef: RefObject<HTMLElement[]>, contain: boolea
     };
 
     let onFocus = (e) => {
-      // If the focused element is in the current scope, and not in the active scope,
-      // update the active scope to point to this scope.
-      let isInScope = isElementInScope(e.target, scopeRef.current);
-      if (isInScope && (!activeScope || !isElementInScope(e.target, activeScope.current))) {
+      // If a focus event occurs outside the active scope (e.g. user tabs from browser location bar),
+      // restore focus to the previously focused node or the first tabbable element in the active scope.
+      let isInAnyScope = isElementInAnyScope(e.target, scopes);
+      if (!isInAnyScope) {
+        if (focusedNode.current) {
+          focusedNode.current.focus();
+        } else if (activeScope) {
+          focusFirstInScope(activeScope.current);
+        } 
+      } else {
+        e.stopPropagation();
         activeScope = scopeRef;
-      }
-
-      // Save the currently focused node in this scope
-      if (isInScope) {
         focusedNode.current = e.target;
       }
+    };
 
-      // If a focus event occurs outside the active scope (e.g. user tabs from browser location bar),
-      // restore focus to the previously focused node or the first tabbable element if none.
-      if (activeScope === scopeRef && !isInScope) {
-        if (focusedNode.current) {
-          focusedNode.current.focus();
-        } else {
-          focusFirstInScope(scopeRef.current);
-        }
+    let onBlur = (e) => {
+      e.stopPropagation();
+      let isInAnyScope = isElementInAnyScope(e.relatedTarget, scopes);
+
+      if (!isInAnyScope) {
+        activeScope = scopeRef;
+        focusedNode.current = e.target;
+        focusedNode.current.focus();
       }
     };
 
     document.addEventListener('keydown', onKeyDown, false);
     document.addEventListener('focusin', onFocus, false);
+    scope.forEach(element => element.addEventListener('focusin', onFocus, false));
+    scope.forEach(element => element.addEventListener('focusout', onBlur, false));
     return () => {
       document.removeEventListener('keydown', onKeyDown, false);
       document.removeEventListener('focusin', onFocus, false);
+      scope.forEach(element => element.removeEventListener('focusin', onFocus, false));
```

Comment:
just to make sure, but the element's we are looping over, they can't change between effects right? I ask because scope is from a ref, which can change without causing `useEffect`

Reply:
A lint error previously mentioned this possibility when I had `scopeRef.current` here previously, and it suggested that I do `let scope = scopeRef.current;` on line 158 so I think we are good here

Reply:
o, yep, that'll do the trick.
hmm... though is there any chance that elements might be null now? so maybe we should `element && element.remove...`?

Reply:
Think we are ok, tested by adding a div and a button that removes said div inside the Dialog's FocusScope. Removing the div and then closing the dialog didn't trigger any console errors, `scope` preserved all 3 element (div, button, and Dialog `section`) while scopeRef.current updated correctly to show button and section only